task/TV3-87: Compress/Extract Toolbar Utility Apps

rstijerina opened the pull request:
## Overview

Enables toolbar utility apps for Tapis v3 compress/extract apps.

## Related

* [TV3-87](https://jira.tacc.utexas.edu/browse/TV3-87)

## Testing

1. Go to any My Data system at https://cep.test/workbench/data and use the compress or extract utility apps. Confirm output appears "in-place", meaning the resulting compressed/uncompressed file(s) appear wherever the original file acted upon was located.

## TODO

- [x] Fix compress app archiving
- [x] ~Migrate these util apps from BATCH to FORK~ Not doing, as FORK seems to not be working at the moment
- [x] Fix tests


For future consideration: Use [file input arrays](https://tapis.readthedocs.io/en/latest/technical/jobs.html#fileinputarrays) for compress

rstijerina commented on client/src/components/Applications/AppForm/AppForm.jsx:
```diff
@@ -273,7 +263,17 @@ export const AppSchemaForm = ({ app }) => {
   };
   let missingAllocation = false;
-  if (app.exec_sys.batchScheduler === 'SLURM') {
+  if (app.definition.jobType === 'BATCH') {
```
Only process queue information if the jobType is `BATCH`

shayanaijaz reviewed (changes requested):
A few things I noticed from testing:
- The correct Compressed File Name doesn't appear after doing a file compress. The compressed file shows up with the name `zip.zip`
- Selecting `.tar.gz` doesn't seem to have an effect. The compressed file still shows up as `zip.zip`
- Compress job fails for files with spaces in them (I realize you have already brought this issue up with cic)
- After an extract, would it make sense to do a check to see if the file already exists and if it does then increment the new file name (e.g in(1).txt) instead of overwriting the file?

nathanfranklin commented on client/src/redux/sagas/datafiles.sagas.js:
```diff
@@ -879,86 +880,72 @@ export function* emptyFile(system, path) {
+            description:
+              'The TACC allocation associated with this job execution',
+            include: true,
+            arg: '-A TACC-ACI',
```
this is related to the cannot-use-FORK-yet issue, right?  Do we need a TODOv3 listed here and/or follow on ticket?  and do we want to pick an allocation from the user's allocations instead of always TACC-ACI?

rstijerina replied:
Oops good catch, that should be a generic allocation. Yeah I'll add a follow on ticket for a FORK compress/extract, but it may turn out all our apps need to run through Slurm in order to get singularity working on some of these hosts

rstijerina replied:
@nathanfranklin Added an allocation selector

rstijerina commented:
@shayanaijaz 
> * The correct Compressed File Name doesn't appear after doing a file compress. The compressed file shows up with the name `zip.zip`
> * Selecting `.tar.gz` doesn't seem to have an effect. The compressed file still shows up as `zip.zip`

Fixed. Can you try again?

> * Compress job fails for files with spaces in them (I realize you have already brought this issue up with cic)

Noticed it fails for any special character as well

> * After an extract, would it make sense to do a check to see if the file already exists and if it does then increment the new file name (e.g in(1).txt) instead of overwriting the file?

I agree it would be nice. That may be overly complicated, though. I think I'm going to instead unpack the items into a dir with the same name as the archive file, with "extract-$timestamp" as well

shayanaijaz commented:
> @shayanaijaz
> 
> > * The correct Compressed File Name doesn't appear after doing a file compress. The compressed file shows up with the name `zip.zip`
> > * Selecting `.tar.gz` doesn't seem to have an effect. The compressed file still shows up as `zip.zip`
> 
> Fixed. Can you try again?


@rstijerina The functionality is still a bit weird. The name shows up correctly but the file extension doesn't seem to be getting set properly. See below screenshot where I entered the Compressed File Name as `Test`

<img width="1553" alt="Screen Shot 2023-03-20 at 4 45 32 PM" src="https://user-images.githubusercontent.com/23081932/226473141-5f72b033-71cb-4e07-a639-d3cda81c4cee.png">

rstijerina commented:
> > @shayanaijaz
> > > * The correct Compressed File Name doesn't appear after doing a file compress. The compressed file shows up with the name `zip.zip`
> > > * Selecting `.tar.gz` doesn't seem to have an effect. The compressed file still shows up as `zip.zip`
> > 
> > 
> > Fixed. Can you try again?
> 
> @rstijerina The functionality is still a bit weird. The name shows up correctly but the file extension doesn't seem to be getting set properly. See below screenshot where I entered the Compressed File Name as `Test`
> 
> <img alt="Screen Shot 2023-03-20 at 4 45 32 PM" width="1553" src="https://user-images.githubusercontent.com/23081932/226473141-5f72b033-71cb-4e07-a639-d3cda81c4cee.png">

@shayanaijaz Thanks, fixed that bug. Also updated extract.

nathanfranklin commented:
Looks good. I'm seeing the following four things: 
*   Archiving a file like `.bash_history(1)(1)`  I get the following error:  /tmp/slurmd/job5336934/slurm_script: line 23: syntax error near unexpected token `(`
* `archiveOnAppError` is false for compress, is that desired?.  It gives you an output location on scratch which is kinda strange, right?  as eventually that will be gone, right?
* compressing `diff.txt` with zip turns into `diff.txt.zip.zip`
* compressing `foo.txt` with tar.gz becomes `foo.txt.tar.gz.zip`

rstijerina commented:
> Looks good. I'm seeing the following four things:
> 
> * Archiving a file like `.bash_history(1)(1)`  I get the following error:  /tmp/slurmd/job5336934/slurm_script: line 23: syntax error near unexpected token `(`

Yeah, I've also seen issues with any non-standard characters. Hoping to find a fix, but probably outside of this PR.

> * `archiveOnAppError` is false for compress, is that desired?.  It gives you an output location on scratch which is kinda strange, right?  as eventually that will be gone, right?

This is desired. Archiving successfully archives to the parent dir of the supplied file for these util apps if the job succeeds. If it fails I want to prevent polluting that space with error logs. The output location is calculated dynamically - could be the job is still running or didn't proceed to the archive phase. Do you have a job id?

> * compressing `diff.txt` with zip turns into `diff.txt.zip.zip`
> * compressing `foo.txt` with tar.gz becomes `foo.txt.tar.gz.zip`

@nathanfranklin Are you for sure pulling the latest from this PR branch?

rstijerina commented:
> > * compressing `diff.txt` with zip turns into `diff.txt.zip.zip`
> > * compressing `foo.txt` with tar.gz becomes `foo.txt.tar.gz.zip`
> 
> @nathanfranklin Are you for sure pulling the latest from this PR branch?

Oh I see the bug.

nathanfranklin commented:
> > Looks good. I'm seeing the following four things:
> > 
> > * Archiving a file like `.bash_history(1)(1)`  I get the following error:  /tmp/slurmd/job5336934/slurm_script: line 23: syntax error near unexpected token `(`
> 
> Yeah, I've also seen issues with any non-standard characters. Hoping to find a fix, but probably outside of this PR.
> 
> > * `archiveOnAppError` is false for compress, is that desired?.  It gives you an output location on scratch which is kinda strange, right?  as eventually that will be gone, right?
> 
> This is desired. Archiving successfully archives to the parent dir of the supplied file for these util apps if the job succeeds. If it fails I want to prevent polluting that space with error logs. The output location is calculated dynamically - could be the job is still running or didn't proceed to the archive phase. Do you have a job id?

Yeah, it was the one with the special chars (`.bash_history(1)(1)`).  So all good if that is a known-issue.

nathanfranklin reviewed (approved):
LGTM 👍 

~Note: Just check tests before merging~. Ah, just codecov failing 🤷‍♂️ 

> Yeah, I've also seen issues with any non-standard characters. Hoping to find a fix, but probably outside of this PR.

and do we need a Jira issue for this?

rstijerina commented:
> LGTM 👍
> 
> ~Note: Just check tests before merging~. Ah, just codecov failing 🤷‍♂️
> 
> > Yeah, I've also seen issues with any non-standard characters. Hoping to find a fix, but probably outside of this PR.
> 
> and do we need a Jira issue for this?

https://jira.tacc.utexas.edu/browse/TV3-145

shayanaijaz reviewed (approved):
LGTM! Functionality working smoothly
